android/app/src/main/java/com/zulipmobile/TextCompressionModule.java

```diff
@@ -0,0 +1,81 @@
+package com.zulipmobile;
+
+import android.util.Base64;
+
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.Promise;
+
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+
+import android.util.Log;
+
+class TextCompressionModule extends ReactContextBaseJavaModule {
+  public TextCompressionModule(ReactApplicationContext reactContext) {
+    super(reactContext);
+  }
+
+  @Override
+  public String getName() {
+    return "TextCompressionModule";
+  }
+
+  // TODO: Experiment what value gives the best performance.
+  private int bufferSize = 8192;
+
+  @ReactMethod
+  public void compress(String input, Promise promise) {
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      Deflater deflater = new Deflater();
+      deflater.setInput(input.getBytes("UTF-8"));
+      deflater.finish();
+      byte[] buffer = new byte[bufferSize];
+      while (!deflater.finished()) {
+        int byteCount = deflater.deflate(buffer);
+        outputStream.write(buffer, 0, byteCount);
+      }
+      deflater.end();
+      outputStream.close();
+      // The RN bridge currently doesn't support sending binary data, so send the compressed
+      // output as a string. Also, out of unknown reasons, using ISO-8859-1 to directly encode
+      // the output as a string doesn't work. Encoding it as a Base64 string does work, but it
+      // is unclear why. Curiously, Base64.encodeToString itself uses ISO-8859-1.
+      promise.resolve(Base64.encodeToString(outputStream.toByteArray(), Base64.DEFAULT));
+    } catch (UnsupportedEncodingException e) {
+      promise.reject("UNSUPPORTED_ENCODING_EXCEPTION", e);
+    } catch (IOException e) {
+      promise.reject("IO_EXCEPTION", e);
+    }
+  }
+
+  @ReactMethod
+  public void decompress(String input, Promise promise) {
+    try {
+      Inflater inflater = new Inflater();
+      inflater.setInput(Base64.decode(input.getBytes("ISO-8859-1"), Base64.DEFAULT));
```

Comment:
Hrmm, okay. But does it make reasoning really easier for you? I find it super helpful to be reassured by this encoding that it'll just work with any bytes that get thrown at it, so I don't need to worry about it at all. For ASCII, I first need to know that the bytes generated by Base64 happen to make up a valid ASCII string.

Reply:
> returns a string that is constructed with the ISO-8859-1 charset

Hmm, what do you mean -- is there something in the chain there that describes it as ISO-8859-1? If so, I agree it could make more sense to match that description.

The actual string doesn't have any charset at all; it consists of a sequence of characters, i.e. Unicode codepoints, not of any particular encoding of those characters into bytes.

Reply:
> Hmm, what do you mean -- is there something in the chain there that describes it as ISO-8859-1?

Nothing in the chain, but rather Base64 itself: https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Encoder.html#encodeToString-byte:A-

> This method first encodes all input bytes into a base64 encoded byte array and then constructs a new String by using the encoded byte array and the ISO-8859-1 charset.

That quote is also what I tried to express with "returns a string that is constructed with the ISO-8859-1 charset".

Reply:
> But does it make reasoning really easier for you? I find it super helpful to be reassured by this encoding that it'll just work with any bytes that get thrown at it, so I don't need to worry about it at all. For ASCII, I first need to know that the bytes generated by Base64 happen to make up a valid ASCII string.

Well, I think we found with a previous experiment you tried that encoding with iso-8859-1, in the context of the APIs handy here, didn't just work in that way. :)

More generally, ASCII is the set of characters where you can always (*) plug your ears and pretend there's just one true character encoding, Unicode was never needed, and the world is simple -- and things will work out just fine. So that's why when I know something is ASCII, it makes it easy to reason about what the code will do. Nothing else works quite so well for blissful ignorance, not even iso-8859-1.

(*) almost -- very occasionally you might run into EBCDIC.

Reply:
> Well, I think we found with a previous experiment you tried that encoding with iso-8859-1, in the context of the APIs handy here, didn't just work in that way. :)

Yeah, that makes sense as well. What's your final call?